libvast/builtins/connectors/file.cpp

```diff
@@ -162,6 +161,9 @@ class plugin : public virtual loader_plugin {
             }
           }
         }
+        if (fd != STDIN_FILENO) {
+          ::close(fd);
```

Comment:
I'm not so sure about this dodgy file descriptor closing.

Reply:
I think this has to be RAII (because upstream generators can just be dropped). Any other reasons for why this might be dodgy?

Reply:
I think this has to be RAII (because upstream generators can just be dropped). Any other reasons for why this might be dodgy?

Reply:
I agree, I don't want to manually handle file descriptors at all, at best.

I could create a struct with custom destructor or a `unique_ptr` with custom destruction function that both wrap the file descriptor and then `close` it upon destruction.

But most of all I'm a bit worried about creation/destruction time. Right now the file descriptor gets created before the generator gets created, but then destroyed inside the generator. This doesn't sit right with me, because right now I would have to `move` that wrapper object into the `std::invoke` call for the generator. Does that make sense to you?

Reply:
Yes, I see what you mean. You could either move the file opening into the beginning of the generator itself (this would also make `load_operator::infer_type_impl` unnecessary, but we can just keep it), but moving the RAII wrapper into the generator seems good to me.

Reply:
I already looked into `file.hpp` and `posix.hpp` - `file` seems like a good abstraction, but right now only for regular files. `posix` has `uds` connection methods that are already used. The main issue that I see here is that `fdinbuf` in the generator only needs and uses a file descriptor to read data - basically no abstraction. Or an abstraction of all abstractions?

Reply:
I opted for a `unique_ptr` with custom deleters in https://github.com/tenzir/vast/pull/3088/commits/26c17b32e7ba53a27f5027a6ab918f74b40d215c

Reply:
This also calls `close(*fd)` if the file could not be opened but this is fine for now.

Reply:
Added `fd` check in https://github.com/tenzir/vast/pull/3088/commits/c46fce7221b2e30b5ca1819e94496a60766aa633